osbenchmark/sync_connection.py

```diff
@@ -0,0 +1,91 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+# The OpenSearch Contributors require contributions made to
+# this file be licensed under the Apache-2.0 license or a
+# compatible open source license.
+# Modifications Copyright OpenSearch Contributors. See
+# GitHub history for details.
+
+import opensearchpy
+
+
+class RequestsHttpConnection(opensearchpy.RequestsHttpConnection):
+    def __init__(self,
+                host= "localhost",
+                port= None,
+                http_auth= None,
+                use_ssl= False,
+                ssl_assert_fingerprint=None,
+                pool_maxsize=None,
+                headers = None,
+                ssl_context = None,
+                http_compress = None,
+                opaque_id = None,
+                **kwargs,):
+        super().__init__(host=host,
+                         port=port,
+                         http_auth=http_auth,
+                         use_ssl=use_ssl,
+                         ssl_assert_fingerprint=ssl_assert_fingerprint,
+                         pool_maxsize=max(256, kwargs.get("max_connections", 0)),
+                         headers=headers,
+                         ssl_context=ssl_context,
+                         http_compress=http_compress,
+                         opaque_id=opaque_id,
+                         **kwargs,)
+
+    def perform_request(self, method, url, params=None, body=None, timeout=None, allow_redirects=True, ignore=(), headers=None):
+        try:
+            # pylint: disable=import-outside-toplevel
+            from osbenchmark.client import RequestContextHolder
+            request_context_holder = RequestContextHolder()
+            request_context_holder.on_request_start()
+            status, headers, raw_data = super().perform_request(method=method, url=url, params=params, body=body, timeout=timeout,
+                                                                allow_redirects=allow_redirects, ignore=ignore, headers=headers)
+            request_context_holder.on_request_end()
```

Comment:
There is no error handling here.  If the request fails, the `on_request_end` callback will not be executed, leading to incorrect metrics in the command dispatcher, or possibly an error there.

Reply:
Note: this may be the wrong location to insert these calls, if they are intended to measure service time.  They should be at the lowest level possible just prior to the HTTP socket send call.  The aiohttp library has callbacks, but the others might not.  In that case, the calls will need to be somewhere within `opensearchpy`, not here.

Reply:
Hey @dblock, any ideas on how we can get the exact timing? Should we make changes in opensearch-py?

Reply:
@dblock

Reply:
For the issue that @gkamat is bringing up, the calling code and the implementation should become a `with` block / `__enter__` and `__exit__` so that `on_request_end` is always called, even on error.

On the second issue of lowest level I think we _are_ trying to measure "client overhead" that begins "around" the client. I also do think it would be good to get lower level HTTPs library metrics for pure transport. Finally, the entire client will become swappable, so keeping the code here makes sense to me.

@gkamat WDYT?

Reply:
@dblock, yes, either a context manager or exception handling will be an appropriate mechanism to handle the issue.

Regarding the second point, I believe those calls were added by @saimedhi to measure _service time_, which reflects the time the request is placed on the wire and received from the network (as accurately as is feasible).  The `aiohttp` client provides trace event callbacks that make this possible.  

You are correct about the client processing time in that it measures latency around the client and that location above would be the correct entry point to measure this metric.  But I believe it is the other metric that is being captured here and the difference between the two would be pure client overhead.